README.md
# Salary Calc
Salary Calc is a tool developed using `Python` that simplifies the process of calculating total salary based on base pay and overtime hours.
The tool enables straightforward and consistent calculations of take-home pay, including any overtime premiums or other relevant compensation.


gpt/changes.txt
now it works. But could you help me change to code with the following changes: 

1. Change just the input dateformat to 1506 for 15th of june and keep 15/06 when displaying the dates.
2. Change so that i should enter time like 1700-2345 as opposed to 17:00-23:45.
3. Display the total accumulated salary after every run, i.e. print the salary.txt file or extract the content etc.
4. Insert a newline between the total amount line and salary entries.


salary_calc/per.py
import datetime

# Base salary
s = float(152.5)

# Overtime scheme
overtime_scheme = {
    "Monday": [
        {"start": "06:00", "end": "18:15", "multiplier": 1},
        {"start": "18:15", "end": "20:00", "multiplier": 1.5},
        {"start": "20:00", "end": "23:30", "multiplier": 1.7},
    ],
    "Tuesday": [
        {"start": "06:00", "end": "18:15", "multiplier": 1},
        {"start": "18:15", "end": "20:00", "multiplier": 1.5},
        {"start": "20:00", "end": "23:30", "multiplier": 1.7},
    ],
    "Wednesday": [
        {"start": "06:00", "end": "18:15", "multiplier": 1},
        {"start": "18:15", "end": "20:00", "multiplier": 1.5},
        {"start": "20:00", "end": "23:30", "multiplier": 1.7},
    ],
    "Thursday": [
        {"start": "06:00", "end": "18:15", "multiplier": 1},
        {"start": "18:15", "end": "20:00", "multiplier": 1.5},
        {"start": "20:00", "end": "23:30", "multiplier": 1.7},
    ],
    "Friday": [
        {"start": "06:00", "end": "18:15", "multiplier": 1},
        {"start": "18:15", "end": "20:00", "multiplier": 1.5},
        {"start": "20:00", "end": "23:30", "multiplier": 1.7},
    ],
    "Saturday": [
        {"start": "06:00", "end": "12:00", "multiplier": 1},
        {"start": "12:00", "end": "23:00", "multiplier": 2},
    ],
    "Sunday": [
        {"start": "06:00", "end": "23:00", "multiplier": 2},
    ],
}

# Get today's date
today = datetime.datetime.now().strftime("%d%m")

# Prompt user for working hours
working_hours = input("Enter your working hours for today (format: 1005-1900), or type 'skip' if you haven't worked today: ")

if working_hours.lower() == "skip":
    print("You have not worked today.")
else:
    # Calculate salary
    start_time = datetime.datetime.strptime(working_hours[:4], "%H%M").time()
    end_time = datetime.datetime.strptime(working_hours[5:], "%H%M").time()
    day_of_week = datetime.datetime.today().strftime("%A")
    salary = 0
    time_slots_salary = []
    for slot in overtime_scheme[day_of_week]:
        slot_start_time = datetime.datetime.strptime(slot["start"], "%H:%M").time()
        slot_end_time = datetime.datetime.strptime(slot["end"], "%H:%M").time()
        if start_time <= slot_start_time and end_time >= slot_end_time:
            duration = datetime.datetime.combine(datetime.date.today(), slot_end_time) - datetime.datetime.combine(datetime.date.today(), slot_start_time)
            salary += duration.seconds / 3600 * s * slot["multiplier"]
            time_slots_salary.append(f"{slot['start']}-{slot['end']}: {duration.seconds / 3600 * s * slot['multiplier']:.2f} SEK")
        elif start_time >= slot_start_time and end_time <= slot_end_time:
            duration = datetime.datetime.combine(datetime.date.today(), end_time) - datetime.datetime.combine(datetime.date.today(), start_time)
            salary += duration.seconds / 3600 * s * slot["multiplier"]
            time_slots_salary.append(f"{working_hours}: {duration.seconds / 3600 * s * slot['multiplier']:.2f} SEK")
        elif start_time >= slot_start_time and start_time <= slot_end_time:
            duration = datetime.datetime.combine(datetime.date.today(), slot_end_time) - datetime.datetime.combine(datetime.date.today(), start_time)
            salary += duration.seconds / 3600 * s * slot["multiplier"]
            time_slots_salary.append(f"{working_hours[:4]}-{slot['end']}: {duration.seconds / 3600 * s * slot['multiplier']:.2f} SEK")
        elif end_time >= slot_start_time and end_time <= slot_end_time:
            duration = datetime.datetime.combine(datetime.date.today(), end_time) - datetime.datetime.combine(datetime.date.today(), slot_start_time)
            salary += duration.seconds / 3600 * s * slot["multiplier"]
            time_slots_salary.append(f"{slot['start']}-{working_hours[5:]}: {duration.seconds / 3600 * s * slot['multiplier']:.2f} SEK")

    # Display salary and time slots salary
    print(f"Your salary for today is: {salary:.2f} SEK")
    print("Time slots salary:")
    for slot_salary in time_slots_salary:
        print(slot_salary)

    # Save salary to file
    with open(f"{today}_salary.txt", "a") as f:
        f.write(f"{today}: {salary:.2f} SEK\n")

    # Calculate and display grand total salary
    grand_total_salary = 0
    with open(f"{today}_salary.txt", "r") as f:
        for line in f:
            grand_total_salary += float(line.split(": ")[1].split(" ")[0])
    print(f"\nGrand total salary: {grand_total_salary:.2f} SEK")


salary_calc/phi.py
from datetime import datetime, timedelta


def parse_time(time_str):
    start, end = time_str.split("-")
    start_time = datetime.strptime(start, "%H%M")
    end_time = datetime.strptime(end, "%H%M")
    return start_time, end_time


def calculate_salary(start_time, end_time, base_salary, day_of_week):
    salary = 0
    current_time = start_time
    
    while current_time < end_time:
        if day_of_week in range(0, 5):  # Monday-Friday
            if current_time.time() < datetime.strptime("18:15", "%H:%M").time():
                salary += base_salary
            elif current_time.time() < datetime.strptime("20:00", "%H:%M").time():
                salary += 1.5 * base_salary
            else:
                salary += 1.7 * base_salary
        elif day_of_week == 5:  # Saturday
            if current_time.time() < datetime.strptime("12:00", "%H:%M").time():
                salary += base_salary
            else:
                salary += 2 * base_salary
        else:  # Sunday
            salary += 2 * base_salary

        current_time += timedelta(minutes=1)
    
    return salary

def save_to_file(date, daily_salary, total_salary):
    with open("salary.txt", "a") as f:
        f.write(f"{date}: {daily_salary:.2f} SEK\n")
        f.write("\n")
        f.write(f"Total Salary: {total_salary:.2f} SEK\n")


def main():
    date = datetime.now().strftime("%d%m")
    input_str = input("Enter working hours in format '1005-1900' or 'skip': ")
    
    if input_str.lower() != "skip":
        start_time, end_time = parse_time(input_str)
        day_of_week = datetime.now().weekday()
        
        base_salary = 152  # You can set your base salary here
        daily_salary = calculate_salary(start_time, end_time, base_salary, day_of_week)
        
        print(f"Total salary for {date}: {daily_salary:.2f} SEK")
        
        # Read previous total salary from the file
        try:
            with open("salary.txt", "r") as f:
                lines = f.readlines()
                previous_total = float(lines[-1].split()[-2])
        except (FileNotFoundError, IndexError, ValueError):
            previous_total = 0
        
        total_salary = previous_total + daily_salary
        save_to_file(date, daily_salary, total_salary)
    else:
        print("Skipped for today.")

if __name__ == "__main__":
    main()


salary_calc/main.py
import datetime
import os
from tabulate import tabulate

# Base salary
s = float(152.5)
# s = float(input("What is your base salary (s) in SEK? "))

# Overtime scheme
overtime_scheme = {
    "Monday": [
        {"start": "06:00", "end": "18:15", "multiplier": 1},
        {"start": "18:15", "end": "20:00", "multiplier": 1.5},
        {"start": "20:00", "end": "23:30", "multiplier": 1.7},
    ],
    "Tuesday": [
        {"start": "06:00", "end": "18:15", "multiplier": 1},
        {"start": "18:15", "end": "20:00", "multiplier": 1.5},
        {"start": "20:00", "end": "23:30", "multiplier": 1.7},
    ],
    "Wednesday": [
        {"start": "06:00", "end": "18:15", "multiplier": 1},
        {"start": "18:15", "end": "20:00", "multiplier": 1.5},
        {"start": "20:00", "end": "23:30", "multiplier": 1.7},
    ],
    "Thursday": [
        {"start": "06:00", "end": "18:15", "multiplier": 1},
        {"start": "18:15", "end": "20:00", "multiplier": 1.5},
        {"start": "20:00", "end": "23:30", "multiplier": 1.7},
    ],
    "Friday": [
        {"start": "06:00", "end": "18:15", "multiplier": 1},
        {"start": "18:15", "end": "20:00", "multiplier": 1.5},
        {"start": "20:00", "end": "23:30", "multiplier": 1.7},
    ],
    "Saturday": [
        {"start": "06:00", "end": "12:00", "multiplier": 1},
        {"start": "12:00", "end": "23:00", "multiplier": 2},
    ],
    "Sunday": [
        {"start": "06:00", "end": "23:00", "multiplier": 2},
    ],
}

# Prompt user for date
date_input = input("Enter the date of the hours you are about to log (format: 1906), or type 'today' if you want to log hours for today: ")
if date_input.lower() == "today":
    today = datetime.datetime.now().strftime("%d%m")
else:
    today = datetime.datetime.strptime(date_input, "%d%m").strftime("%d%m")

# Prompt user for working hours
working_hours = input("Enter your working hours for today (format: 1005-1900), or type 'skip' if you haven't worked today: ")

if working_hours.lower() == "skip":
    print("You have not worked today.")
else:
    # Calculate salary
    start_time = datetime.datetime.strptime(working_hours[:4], "%H%M").time()
    end_time = datetime.datetime.strptime(working_hours[5:], "%H%M").time()
    day_of_week = datetime.datetime.today().strftime("%A")
    salary = 0
    time_slots_salary = []
    for slot in overtime_scheme[day_of_week]:
        slot_start_time = datetime.datetime.strptime(slot["start"], "%H:%M").time()
        slot_end_time = datetime.datetime.strptime(slot["end"], "%H:%M").time()
        if start_time <= slot_start_time and end_time >= slot_end_time:
            duration = datetime.datetime.combine(datetime.date.today(), slot_end_time) - datetime.datetime.combine(datetime.date.today(), slot_start_time)
            salary += duration.seconds / 3600 * s * slot["multiplier"]
            time_slots_salary.append(f"{slot['start']}-{slot['end']}: {duration.seconds / 3600 * s * slot['multiplier']:.2f} SEK")
        elif start_time >= slot_start_time and end_time <= slot_end_time:
            duration = datetime.datetime.combine(datetime.date.today(), end_time) - datetime.datetime.combine(datetime.date.today(), start_time)
            salary += duration.seconds / 3600 * s * slot["multiplier"]
            time_slots_salary.append(f"{working_hours}: {duration.seconds / 3600 * s * slot['multiplier']:.2f} SEK")
        elif start_time >= slot_start_time and start_time <= slot_end_time:
            duration = datetime.datetime.combine(datetime.date.today(), slot_end_time) - datetime.datetime.combine(datetime.date.today(), start_time)
            salary += duration.seconds / 3600 * s * slot["multiplier"]
            time_slots_salary.append(f"{working_hours[:4]}-{slot['end']}: {duration.seconds / 3600 * s * slot['multiplier']:.2f} SEK")
        elif end_time >= slot_start_time and end_time <= slot_end_time:
            duration = datetime.datetime.combine(datetime.date.today(), end_time) - datetime.datetime.combine(datetime.date.today(), slot_start_time)
            salary += duration.seconds / 3600 * s * slot["multiplier"]
            time_slots_salary.append(f"{slot['start']}-{working_hours[5:]}: {duration.seconds / 3600 * s * slot['multiplier']:.2f} SEK")

# ... (all the code above remains unchanged)

# Initialize table_data and headers
    table_data = []
    headers = ["Date", "Working hours", "Salary"]

# Display salary and time slots salary
    print(f"Your salary for {today[:2]}/{today[2:]} is: {salary:.2f} SEK")
    print("Time slots salary:")
    for slot_salary in time_slots_salary:
        print(slot_salary)

# Calculate grand total salary
    grand_total_salary = 0  # Initialize grand_total_salary to 0

# Read the content of salaries.txt
    with open("salaries.txt", "r") as f:
        lines = f.readlines()

# Update the table_data list and grand_total_salary
    for line in lines:
        if "|" in line and "Date" not in line and "Grand total salary" not in line and "-------" not in line:
            date_value = line.split("|")[1].strip()
            working_hours_value = line.split("|")[2].strip()
            salary_value = line.split("|")[3].strip()
            table_data.append([date_value, working_hours_value, salary_value])
            try:
                grand_total_salary += float(salary_value.split(" ")[0])
            except (ValueError, IndexError):
                # Skip the line if it cannot be converted to a float or has an incorrect index
                pass

# Add the current day's salary to table_data and grand_total_salary
    start_time, end_time = working_hours.split("-")
    formatted_working_hours = f"{start_time[:2]}:{start_time[2:]}-{end_time[:2]}:{end_time[2:]}"
    table_data.append([f"{today[:2]}/{today[2:]}", formatted_working_hours, f"{salary:.2f} SEK"])
    grand_total_salary += salary

# Add an empty row before the grand total salary row
    # table_data.append(["", "", ""])

# Add the grand total salary row to the table_data list
    table_data.append(["", "Grand total salary:", f"{grand_total_salary:.2f} SEK"])

# Save the updated table_data to salaries.txt
    with open("salaries.txt", "w") as f:
        f.write(tabulate(table_data, headers=headers, tablefmt="pipe"))

# Display the table in the terminal using the tabulate library
    print(tabulate(table_data, headers=headers, tablefmt="pipe"))
